Add create tenant from CLI procedure (#500)
minio/docs: source/operations/install-deploy-manage/deploy-minio-tenant.rst

      Ensure all Persistent Volumes provisioned to support the MinIO Tenant 
      use this storage class.

.. _minio-k8s-deploy-minio-tenant-commandline:

Procedure (MinIO Operator Console)
----------------------------------

> To deploy a tenant from the MinIO Operator Console, complete the following steps:

Reviewer: ```suggestion
To deploy a tenant from the MinIO Operator Console, complete the following steps in order:
```

Author: 👍

      Ensure all Persistent Volumes provisioned to support the MinIO Tenant 
      use this storage class.

.. _minio-k8s-deploy-minio-tenant-commandline:

Procedure
---------

> .. tab-set::

Reviewer: So I'm looking at this, and I'm a little unsure if we should use tabs for these procedures:

1. We lose the header anchors for individual steps, and along with that the ability to link to specific sections
2. We imply that Operator Console = CLI in functionality, which is not the case. The Operator Console has far more flexibility
3. Eventually we need to add `Kustomize` or something for very advanced users, which I guess is not a big deal having three tabs.

The gain in not scrolling is kind of nice - but we have a floating right TOC, so a user could click to whatever H1 they want.

Thus my proposal - we should have three H1s on this page:

- Deploy MinIO Tenant using Operator Console
- Deploy MinIO Tenant using `kubectl minio`
- (Future) Deploy MinIO Tenant using `Kustomize`

Each section can have it's own pre-requisites (if necessary).  That also lets us describe specific behaviors cleanly (e.g. w/ `kubectl minio` TLS is always enabled)

LMK your thoughts.

Reviewer: :thinking: there should be a way to get that scroll to work...though to your point if we have that many headings in the right TOC something else needs to get fixed up.

          Applications deployed inside the cluster can access the services using the ``CLUSTER-IP``. 

          Applications external to the Kubernetes cluster can access the services using the ``EXTERNAL-IP``. 
          This value is only populated for Kubernetes clusters configured for Ingress or a similar network access service. 
          Kubernetes provides multiple options for configuring external access to services. 
          See the Kubernetes documentation on 
          :kube-docs:`Publishing Services (ServiceTypes) <concepts/services-networking/service/#publishing-services-service-types>` 
          and :kube-docs:`Ingress <concepts/services-networking/ingress/>` 
          for more complete information on configuring external access to services.

   .. tab-item:: Command Line (CLI)

      The :mc:`kubectl minio tenant create` command supports creating a MinIO Tenant in your Kubernetes cluster.
      The command *requires* that the cluster have a functional MinIO Operator installation.

      You can also use this procedure to generate a Kubernetes ``YAML``-formatted resource document for further customization.

>       Prerequisite

Author: Removed.

        Initialize the MinIO Operator.

>       1) Create a Tenant with default options

Reviewer: So this is an example of how the step-style doesn't really work. These aren't sequential steps - they are parallel options. You can do 1 or 2 or 3, all of which result in 4.

If this was a standalone procedure, *this* is where tabs would make a lot of sense.

Author: When it comes to the CLI, it's really one step: run the `kubectl minio tenant create` command with the required and desired optional flags.

Reworded the steps to reflect a decision process rather than an action process.

   
         Use the :mc:`kubectl minio tenant create` command to add a new tenant.

> 

Author: Moved it up

         - :mc:`~kubectl minio tenant create TENANT_NAME`
     
           The name to use for the new tenant.

         - :mc:`~kubectl minio tenant create --capacity`

           The total raw storage size for the Tenant. 
           Specify both the total storage size *and* the :guilabel:`Unit` of that storage. 
           All storage units are in SI values, e.g. Gi = GiB = 1024\ :sup:`3` bytes.
     
           The total storage capacity for the tenant to have across all volumes.
           The value should be an integer and a standard storage abbreviation, such as ``Gi`` or ``Ti``.

           For example, 16 Ti for 16 Tibibytes.

         - :mc:`~kubectl minio tenant create --servers`

           The total number of MinIO server pods to deploy in the Tenant.
    
           The Operator by default uses pod anti-affinity, such that the Kubernetes cluster *must* have at least one worker node per MinIO server pod.

         - :mc:`~kubectl minio tenant create --volumes`

           The total number of storage volumes (Persistent Volume Claims). 
           The Operator generates an equal number of PVC *plus two* for supporting Tenant services (Metrics and Log Search).
       

>            The specified :guilabel:`Storage Class` *must* correspond to a set of Persistent Volumes sufficient in number to match each generated PVC.

Reviewer: This feels a little weird - the user at this point does not know how to specify this. Maybe sub with a reference to the `--storage-class` arg and lead with "Specify a storage class using <arg>`, or omit to use the `default` Kubernetes storage class.

    
           The total number of persistent volume claims (``PVC``) per server is determined by dividing the number of volumes by the number of servers.
           The storage available for each ``PVC`` is determined by dividing the capacity by the number of volumes. 

           The generated claims have pod selectors so that claims are only made for volumes attached to node running the pod.

           If the number of volumes exceeds the numnber of persistent volumes available on the cluster, ``MinIO`` hangs until the number of persistent volumes are available.
  
         - :mc:`~kubectl minio tenant create --namespace`

           Each MinIO tenant requires its own ``namespace``.

           When not specified, the MinIO Operator defaults to use ``minio``.
           To specify a different namespace, use the :mc:`~kubectl minio tenant create --namespace` flag.

           The namespace must already exist in the Kubernetes cluster.
           Run ``kubectl create ns <new_namespace>`` to add one.

>          For example, the following command creates a new tenant with 16 volumes, total capacity of 16 Tebibytes, using the latest MinIO image, on four servers, in the ``minio`` namespace.

Reviewer: Lot of commas. Might be more readable as a bullet list.

Author: Broke it out into a bulleted list

         .. code-block:: shell
            :class: copyable

            kubectl minio tenant create minio_tenant --capacity 16Ti --servers 4 --volumes 16 --namespace minio

>       2) Add any optional configurations

Reviewer: Too late to do this after running step 1. 

Maybe restructure to *not* run the command in step 1 - have the user "Enter the required configurations for the tenant", then "add optional configurations", then "Deploy the Tenant", etc.

         You can further customize your tenant while executing the command by including any or all of the following *optional* flags.

         - :mc:`~kubectl minio tenant create --image`
           
           Customize the ``minio`` image to use.
  

>            By default, the Operator uses the latest ``minio`` release image when creating the tenant.

Reviewer: Does it? I thought each Operator version had a static release [based on the Operator release](https://github.com/minio/operator/blob/master/pkg/apis/minio.min.io/v2/constants.go#L91). They can specify `minio/minio:latest` to use latest stable.

Author: Good catch.

  
           For more, see the `Github documentation <https://github.com/minio/kes/wiki>`__.

         - :mc:`~kubectl minio tenant create --storage-class`
           
           Specify the storage class to use.

           New MinIO tenants use the ``default`` storage class.
           To specify a different storage class, add the :mc:`~kubectl minio tenant create --storage-class` flag.

           The specified :mc-cmd:`~kubectl minio tenant create --storage-class` *must* match the ``storage-class`` of the Persistent Volumes (``PVs``) to which the ``PVCs`` should bind.

           MinIO strongly recommends creating a Storage Class that corresponds to locally-attached volumes on the host machines on which the Tenant deploys. 
           This ensures each pod can use locally-attached storage for maximum performance and throughput. 
           See the :ref:`Deploy MinIO Tenant <minio-k8s-deploy-minio-tenant>` tutorial for guidance on creating Storage Classes for supporting the MinIO Tenant.

      3) Generate a YAML file for further customizations

>          Other options exist for customizing a MinIO tenant.

Reviewer: https://github.com/minio/operator/blob/master/docs/crd.adoc

So MinIO Operator installs a Custom Resource Definition for describing Tenants. Advanced users can generate a base YAML using the CLI and customize based on the CRD.

Author: I'll have to read up on this. For now, I'll just essentially borrow this language and insert it.

         To customize additional configurations, edit a YAML file.

         Do a dry run of a tenant creation process and generate a YAML file with the command line options with the :mc:`~kubectl minio tenant create --output` flag.

         When using this flag, the operator does **not** create the tenant.
         Modify the generated YAML file as desired, then use ``kubectl apply -f <FILE>`` to manually create the MinIO tenant using the file.

      4) Record the Access Information

         When generating the tenant, the MinIO Operator displays the access credentials to use for the tenant.

         .. important::
   
            This is the only time the credentials display.
            They cannot be retrieved later.

> 

Reviewer: passive voice. Spitballing:

"Copy the displayed credentials to a secure location. MinIO does not display these credentials again."

            Record the username and password to a secure location before closing the terminal.

>          In addition to access credentials, the output shows the service name and service reports to use for accessing the tenant.

Reviewer: "service reports" ?

Author: PORTS. PORTS, dagnabit!

         To access the :ref:`MinIO Console <minio-console>` for the tenant, forward the tenant's port.

         - If necessary, run ``kubectl get svc -n <namespace>`` to retrieve the tenant's port number.
  
         - Run the following to forward the tenant's port and access it from a browser:

            .. code-block:: shell
               :class: copyable

               kubectl port-forward svc/<tenant-name>-console -n <tenant-namespace> <localport>:<tenantport>

           - Replace ``<tenant-name>`` with the name of your tenant.
  
           - Replace ``<tenant-namespace>`` with the namespace the tenant exists in.
  
           - Replace ``<localport>`` with the port number to use on your local machine to access the tenant's MinIO Console.
  
           - Replace ``<tenantport>`` with the port number the MinIO Operator assigned to the tenant.

         - Go to ``https://127.0.0.1:<localport>`` to Access the tenant's MinIO Console.
 
           Replace ``<localport>`` with the port number you used when forwarding the tenant's port.

         - Login with the username and password shown in the tenant creation output and recorded in step 4 above.

> Forward Ports

Reviewer: This feels oddly alone - there's no context on where to use this.

You can just take the instructions and add it to the "Connect" step for each procedure.

Author: Moved to be an additional step on each procedure.

         kubectl port-forward service/minio-tenant-1-console 9443:9443

Procedure (Command Line)
------------------------------

The :mc:`kubectl minio tenant create` command supports creating a MinIO Tenant in your Kubernetes cluster.
The command *requires* that the cluster have a functional MinIO Operator installation.

To deploy a tenant from the command line, complete the following steps:

:ref:`create-tenant-cli-determine-settings-required-options`

:ref:`create-tenant-cli-determine-additional-options`

:ref:`create-tenant-cli-enter-command`

:ref:`create-tenant-cli-record-access-info`

:ref:`create-tenant-cli-access-tenant-console`

:ref:`create-tenant-cli-forward-ports`

.. _create-tenant-cli-determine-settings-required-options:

> 1) Determine Settings for Required Settings

Author: Went with "Determine Values for Required Settings"

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :mc:`kubectl minio tenant create` command requires several configuration settings.

> Determine the values for all required flags.

Reviewer: Or...we could use something like https://docs.min.io/minio/baremetal/reference/minio-mc-admin/mc-admin-user-svcacct.html#mc.admin.user.svcacct.add

and have two tabs - a list of the required settings (as you have below) and an example usage.

> Determine the values for all required flags.

Author: I like the tabbing idea. Going with that.

For example, the following command creates a new tenant with 
  
- the name ``minio_tenant``
- total capacity of 16 Tebibytes
- four servers
- 16 volumes

> - the ``minio`` namespace.

Reviewer: ```suggestion
- in the ``minio`` namespace.
```

Author: That doesn't work the the preposition at the end of the introductory statement.
Switching it up and doing a definition list instead.

.. code-block:: shell
   :class: copyable

   kubectl minio tenant create minio_tenant --capacity 16Ti --servers 4 --volumes 16 --namespace minio

The command requires values for each of the following:

- :mc:`~kubectl minio tenant create TENANT_NAME`
    
  The name to use for the new tenant.

- :mc:`~kubectl minio tenant create --capacity`

  The total raw storage size for the Tenant across all volumes. 
  Specify both the total storage size *and* the :guilabel:`Unit` of that storage. 
  All storage units are in SI values, e.g. :math:`Gi = GiB = 1024^3` bytes.
     
  For example, 16 Ti for 16 Tebibytes.

- :mc:`~kubectl minio tenant create --servers`

  The total number of MinIO server pods to deploy in the Tenant.
    
  The Operator by default uses pod anti-affinity, such that the Kubernetes cluster *must* have at least one worker node per MinIO server pod.

- :mc:`~kubectl minio tenant create --volumes`

  The total number of storage volumes (Persistent Volume Claims). 

>   The Operator generates an equal number of PVC *plus two* for supporting Tenant services (Metrics and Log Search).

Author: It did not create these when I tested. Removed.

>   The Operator generates an equal number of PVC *plus two* for supporting Tenant services (Metrics and Log Search).

Author: It did create one for logs, but not for metrics.

       
  The total number of persistent volume claims (``PVC``) per server is determined by dividing the number of volumes by the number of servers.
  The storage available for each ``PVC`` is determined by dividing the capacity by the number of volumes. 

  The generated claims have pod selectors so that claims are only made for volumes attached to node running the pod.

  If the number of volumes exceeds the numnber of persistent volumes available on the cluster, ``MinIO`` hangs until the number of persistent volumes are available.
  
- :mc:`~kubectl minio tenant create --namespace`

  Each MinIO tenant requires its own ``namespace``.

  When not specified, the MinIO Operator defaults to use ``minio``.

>   To specify a different namespace, use the :mc:`~kubectl minio tenant create --namespace` flag.

Reviewer: Since we're saying these are required, its odd to talk about a default value at all. Maybe

"Omit this value to use the default namespace ``minio``."

Or drop this section entirely and go straight to 622

Author: Did a few updates to maybe address this.

  The namespace must already exist in the Kubernetes cluster.
  Run ``kubectl create ns <new_namespace>`` to add one.

.. _create-tenant-cli-determine-additional-options:

2) Determine Additional Optional Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can further customize your tenant while executing the command by including any or all of the *optional* flags.

Reviewer: ```suggestion
You can further customize your tenant while by including any or all of the following *optional* flags when running the :mc:`kubectl minio tenant create` command.
```

Author: This is better. I also dropped the following line as no longer necessary.

Determine whether to use each of the following and the value to set:

- :mc:`~kubectl minio tenant create --image`
           
  Customize the ``minio`` image to use.
  
  By default, the Operator uses the release image available at the time of the Operator's release.
  To specify a different MinIO version for the tenant, such as the latest available, use the :mc:`~kubectl minio tenant create --image` flag.

  See the `MinIO Quay <https://quay.io/repository/minio/minio>`__ or the `MinIO DockerHub <https://hub.docker.com/r/minio/minio/tags>`__ repositories for a list of valid tags.

- :mc:`~kubectl minio tenant create --image-pull-secret`
           

>   If required, specify the secret to use when pulling the ``minio`` image.

Reviewer: We should specify if at all possible why this would be required. I _think_ in this case it's if the specified image repository requires a password for access.

Author: For custom container registries. Clarified.

  Use :mc:`~kubectl minio tenant create --image-pull-secret` to specify the secret.

- :mc:`~kubectl minio tenant create --kes-config`
           
  Configure a :minio-git:`Key Encrption Service (KES) <kes>`

  Use the :mc:`~kubectl minio tenant create --kes-config` flag to specify the name of the secret to use for KES Key Management Service (KMS) setup.

  Enabling Server Side Encryption (SSE) also deploys a MinIO :minio-git:`KES <kes>` service in the Tenant to faciliate SSE operations.
  
  For more, see the `Github documentation <https://github.com/minio/kes/wiki>`__.

> - :mc:`~kubectl minio tenant create --storage-class`

Reviewer: 🤔 so this is strictly not *optional*, but I think failing to specify this up front is likely to result in failure.  The user has to make sure the specified storage class exists and has sufficient PVs to bind. I'd move this up.

Author: Moved to the required section.

  MinIO strongly recommends creating a Storage Class that corresponds to locally-attached volumes on the host machines on which the Tenant deploys. 
  This ensures each pod can use locally-attached storage for maximum performance and throughput. 
  See the :ref:`Deploy MinIO Tenant <minio-k8s-deploy-minio-tenant>` tutorial for guidance on creating Storage Classes for supporting the MinIO Tenant.

.. note:: Generate a YAML File for Further Customizations

   The MinIO Operator installs a `Custom Resource Definition (CRD) <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/>`__ to describe tenants.
   Advanced users can generate a YAML file from the command line and customize the tenant based on the CRD.

   Do a dry run of a tenant creation process to generate a YAML file using the :mc:`~kubectl minio tenant create --output` flag.

   When using this flag, the operator does **not** create the tenant.
   Modify the generated YAML file as desired, then use ``kubectl apply -f <FILE>`` to manually create the MinIO tenant using the file.

.. _create-tenant-cli-enter-command:

> 3) Enter the Command with Required and Optional Settings

Reviewer: ```suggestion
3) Run the Command with Required and Optional Settings
```

Author: 👍

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

At the command line, enter the full command with all *Required* and any *Optional* flags.

Consider a tenant we want to create:

- The tenant is ``minio1`` and has a total storage capacity of 16 Tebibytes.
- The tenant uses four nodes with four volumes per node, or 16 total volumes.
- The tenant should exist in the ``minio-tenant-space`` namespace.
- The tenant should use the latest MinIO version.
- The key ecryption file is ``minio-secret``.
- The tenant uses the Operator's ``warm`` storage.

.. code-block:: shell
   :substitutions:

   kubectl minio tenant create                                \
                        minio1                                \
                        --capacity 16Ti                       \
                        --servers 4                           \
                        --volumes 16                          \
                        --namespace minio-tenant-space        \
                        --image |minio-latest| \
                        --kes-config minio-secret             \
                        --storage-class warm

.. _create-tenant-cli-record-access-info:

> 4) Record the Access Information

Reviewer: ```suggestion
4) Record the Access Credentials
```

Author: Yep, that's better.

.. _minio-k8s-deploy-minio-tenant-commandline:

Procedure (Command Line)
------------------------

The :mc:`kubectl minio tenant create` command supports creating a MinIO Tenant in your Kubernetes cluster.
The command *requires* that the cluster have a functional MinIO Operator installation.

To deploy a tenant from the command line, complete the following steps:

:ref:`create-tenant-cli-determine-settings-required-options`

:ref:`create-tenant-cli-determine-additional-options`

:ref:`create-tenant-cli-enter-command`

:ref:`create-tenant-cli-record-access-info`

:ref:`create-tenant-cli-access-tenant-console`

:ref:`create-tenant-cli-forward-ports`

.. _create-tenant-cli-determine-settings-required-options:

1) Determine Values for Required Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :mc:`kubectl minio tenant create` command requires several configuration settings.
Determine the values for all required settings.

.. tab-set::
   
   .. tab-item:: Required Settings

      The command requires values for each of the following:

>       - :mc:`~kubectl minio tenant create TENANT_NAME`

Reviewer: This might read cleaner as a list-table directive

```
.. list-table::
   :header-rows: 1
   :widths: 40 60
   :width: 100%

   * - Setting
      - Description

   * - ...
     - ...

```

Similar to the Operator Console sections above.

Author: Changed to a table

   Modify the generated YAML file as desired, then use ``kubectl apply -f <FILE>`` to manually create the MinIO tenant using the file.

.. _create-tenant-cli-enter-command:

3) Run the Command with Required and Optional Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

At the command line, enter the full command with all *Required* and any *Optional* flags.

Consider a tenant we want to create:

Tenant Name
  ``minio1``

Capacity
  16 Tebibytes

Servers
  4

Volumes
  16 (four per node)

Namespace
  ``miniotenantspace``

MinIO Image
  Latest version, |minio-latest|

Key ecnryption file
  ``minio-secret``

Storage class
  ``warm``

.. code-block:: shell
   :substitutions:

   kubectl minio tenant create                                \
                        minio1                                \
                        --capacity 16Ti                       \
                        --servers 4                           \
                        --volumes 16                          \
                        --namespace miniotenantspace          \
                        --image |minio-latest|  \

>                         --kes-config minio-secret             \

Reviewer: ```suggestion
                        --kes-config minio-kes-config             \
```

`minio-secret` might be a default secret name for tenant credentials

Author: Updated the example
